_posts/2023-04-19-#27039.md

```diff
@@ -7,20 +7,47 @@ authors: [pinheadmz]
+
+- A reindex, amongst other things, recreates the block index from the block files on disk. The block files themselves don't need to be updated: blocks are already assumed to not be stored sequentially in or across block files anyway.
+
+- This PR skips flushing to disk those blocks that already have a known position in the datastore. Skipping this means that users can write-protect the blk files on disk which may be useful for security or even safely sharing that data between multiple bitcoind instances.
+
+- Some of the functions that are relevant to the reindexing process behave differently than can perhaps be expected from their name:
+  - `SaveBlockToDisk()` does not actually write anything to disk when (as is the case for a reindex) the known block position is passed (as `dbp`), but we still need to call it to update the blockfile info stats
+  - `FindBlockPos()` is mostly used to find the position in a blockfile where a new block is going to be appended. However, it also updates the blockfile info and may also trigger the flushing of block and undo files, which is why we still call it even if we already know the block's position.
+  - this behaviour arguably could arguably be improved upon, as for example done in [this WIP by mzumsande](https://github.com/bitcoin/bitcoin/compare/master...mzumsande:bitcoin:202207_refactor_findblockpos)
 
 
 ## Questions
 
 1. Did you review the PR? [Concept ACK, approach ACK, tested ACK, or NACK](https://github.com/bitcoin/bitcoin/blob/master/CONTRIBUTING.md#peer-review)? What was your review approach?
--->
+
+1. After this PR, can the whole `blocks/` directory be made read-only? Why (not)?
+
+1. In [`FindBlockPos()`](https://github.com/bitcoin/bitcoin/blob/19764dc143281376ea08e954018479ed10405b72/src/node/blockstorage.cpp#L611), what is the relevance of `fKnown` to this PR, conceptually? How does the behaviour of `FindBlockPos()` change when `fKnown==true` instead of `false`?
+
+1. What happens if bitcoind is killed after when block(s) have been processed, but before `FlushBlockFile()` is called? Does this PR affect that behaviour?
+
+1. In the `blockmanager_flush_block_file` unit test, why are we [pausing](https://github.com/bitcoin-core-review-club/bitcoin/commit/470ef396b5498d8689802c359a216d5a3c4749a5#diff-d6d633592a40f5f3d8b03863e41547de8751b874c1d20f129a616b9dd719b999R170) before starting a new blockfile? Can you think of issues with this approach, or alternatives?
+
+1. In the `blockmanager_flush_block_file` unit test, why do we still [only have two blocks](https://github.com/bitcoin-core-review-club/bitcoin/blob/470ef396b5498d8689802c359a216d5a3c4749a5/src/test/blockmanager_tests.cpp#L153-L154) in the file after having written `block3` to it?
```

Comment:
We *attempt* to write block3 but we know it won't be written. As mzumsande points out in a comment, we would never overwrite a block with different data, I'm just using this bizarre edge case to enforce the test.

Reply:
Re: pinheadmz's comment, maybe this instead?
```suggestion
1. In the `blockmanager_flush_block_file` unit test, why do we still [only have two blocks](https://github.com/bitcoin-core-review-club/bitcoin/blob/470ef396b5498d8689802c359a216d5a3c4749a5/src/test/blockmanager_tests.cpp#L153-L154) in the file after we have called `SaveBlockToDisk(block3)`?
```